runtime/src/commonMain/kotlin/app/cash/sqldelight/Transacter.kt

```diff
@@ -220,8 +290,67 @@ abstract class TransacterImpl(private val driver: SqlDriver) : Transacter {
     return transactionWithWrapper(noEnclosing, bodyWithReturn)
   }
 
+  override suspend fun <R> suspendingTransactionWithResult(
+    noEnclosing: Boolean,
+    bodyWithReturn: suspend SuspendingTransactionWithReturn<R>.() -> R
+  ): R {
+    return suspendingTransactionWithWrapper(noEnclosing, bodyWithReturn)
+  }
+
+  override suspend fun suspendingTransaction(
+    noEnclosing: Boolean,
+    body: suspend SuspendingTransactionWithoutReturn.() -> Unit
+  ) {
+    return suspendingTransactionWithWrapper(noEnclosing, body)
+  }
+
+  private fun <R> postTransactionCleanup(transaction: Transaction, enclosing: Transaction?, thrownException: Throwable?, returnValue: R?): R {
+    if (enclosing == null) {
+      if (!transaction.successful || !transaction.childrenSuccessful) {
+        // TODO: If this throws, and we threw in [body] then create a composite exception.
+        try {
+          transaction.postRollbackHooks.forEach { it.invoke() }
+        } catch (rollbackException: Throwable) {
+          thrownException?.let {
+            throw Throwable("Exception while rolling back from an exception.\nOriginal exception: $thrownException\nwith cause ${thrownException.cause}\n\nRollback exception: $rollbackException", rollbackException)
+          }
+          throw rollbackException
+        }
+        transaction.postRollbackHooks.clear()
+      } else {
+        if (transaction.pendingTables.isNotEmpty()) {
+          driver.notifyListeners(transaction.pendingTables.toTypedArray())
+        }
+        transaction.pendingTables.clear()
+        transaction.registeredQueries.clear()
+        transaction.postCommitHooks.forEach { it.invoke() }
+        transaction.postCommitHooks.clear()
+      }
+    } else {
+      enclosing.childrenSuccessful = transaction.successful && transaction.childrenSuccessful
+      enclosing.postCommitHooks.addAll(transaction.postCommitHooks)
+      enclosing.postRollbackHooks.addAll(transaction.postRollbackHooks)
+      enclosing.registeredQueries.addAll(transaction.registeredQueries)
+      enclosing.pendingTables.addAll(transaction.pendingTables)
+    }
+
+    if (enclosing == null && thrownException is RollbackException) {
+      // We can safely cast to R here because the rollback exception is always created with the
+      // correct type.
+      @Suppress("UNCHECKED_CAST")
+      return thrownException.value as R
+    } else if (thrownException != null) {
+      throw thrownException
+    } else {
+      // We can safely cast to R here because any code path that led here will have set the
+      // returnValue to the result of the block
+      @Suppress("UNCHECKED_CAST")
+      return returnValue as R
+    }
+  }
+
   private fun <R> transactionWithWrapper(noEnclosing: Boolean, wrapperBody: TransactionWrapper<R>.() -> R): R {
```

Comment:
actually if we wanted to get crazy now that I think about it....three types:

`BaseTransacterImpl` with all the shared stuff
`Transacter` which has `fun transactionWithWrapper` etc.
`SuspendingTransacter` which has `suspend fun transactionWithWrapper`

so if you swap between async/sync...the APIs don't change! the functions just become suspend

Reply:
I like that. I'll make these changes.